File: code/py/muxcat.py
import argparse
import os

# Constants
NSM = b"\xFF\xFE"  # Next Stream Marker
NFM = b"\xFF\xFD"  # Next Frame Marker
NCM = b"\xFF\xFC"  # Next Chunk Marker
COLUMNSIZE = 20

# Argument Parser
parser = argparse.ArgumentParser(description="Parse and display XTXT file streams.")
parser.add_argument("file", help="The XTXT file to process.")
parser.add_argument("-n", "--numbers", action="store_true", help="Display line numbers.")
parser.add_argument("-s", "--stream", type=int, help="Display only the specified stream index.")
parser.add_argument("-w", "--width", type=int, default=COLUMNSIZE, help="Specify column width.")
parser.add_argument("-H", "--head", action="store_true", help="Treat the first line as a header.")
parser.add_argument("-l", "--line", type=int, help="Display only the specified line.")
args = parser.parse_args()

# File Reading
if not os.path.isfile(args.file):
    print(f"Error: File '{args.file}' not found.")
    exit(1)

with open(args.file, "rb") as f:
    txt = f.read()

# Parse Markers
atxt = bytearray(txt)
idx = 0
markers = []

while idx < len(atxt):
    try:
        markpos = atxt.index(0xFF, idx)
        nextbyte = atxt[markpos + 1]

        if nextbyte == NSM[1]:
            markers.append((NSM[1], markpos))
        elif nextbyte == NFM[1]:
            markers.append((NFM[1], markpos))
        elif nextbyte == NCM[1]:
            markers.append((NCM[1], markpos))
        else:
            print(f"Error: Invalid marker at position {markpos}: 0x{nextbyte:02X}")
            exit(1)

        idx = markpos + 2
    except ValueError:
        break

streams = []
stream_widths = []
frame_no = 0
stream_no = 0
idx = 0

# Process Markers
for mtype, mpos in markers:
    aline = atxt[idx:mpos]
    idx = mpos + 2

    if mtype == NSM[1]:
        while len(streams) <= stream_no:
            streams.append([])
            stream_widths.append(0)

        line = aline.decode("utf-8", errors="replace").rstrip()
        streams[stream_no].append(line)
        stream_widths[stream_no] = max(stream_widths[stream_no], len(line))
        stream_no += 1

    elif mtype == NFM[1]:
        stream_no = 0
        frame_no += 1

# Calculate Line Numbers and Widths
longest_stream = max(len(s) for s in streams)
line_no_width = len(str(longest_stream))
line_no_fmt = f"%0{line_no_width}d "

# Output
for line_no in range(longest_stream):
    if args.line and line_no + 1 != args.line:
        continue

    line = ""
    for stream_idx, stream in enumerate(streams):
        if args.stream is not None and stream_idx != args.stream:
            continue

        col_width = args.width or stream_widths[stream_idx]
        content = stream[line_no] if line_no < len(stream) else ""
        line += content.ljust(col_width)

    if args.numbers:
        print(f"{line_no_fmt % (line_no + 1)}{line}")
    else:
        print(line)

File: code/rb/muxcat.rb
#!/usr/bin/env ruby
require 'optparse'
require 'pp'

class String
  def ascii!
    self.force_encoding("ascii-8bit")
  end

  def utf8!
    self.force_encoding("utf-8")
  end
end

# Constants
NSM = "\xFF\xFE".ascii! # Next Stream Marker
NFM = "\xFF\xFD".ascii! # Next Frame Marker
NCM = "\xFF\xFC".ascii! # Next Chunk Marker

# Defaults
COLUMNSIZE = 20
opts = {
  show_line_numbers: true,
  stream_index: nil,
  column_width: COLUMNSIZE,
  treat_as_header: false,
  specific_line: nil
}

# Parse command line options
OptionParser.new do |parser|
  parser.banner = "Usage: muxcat.rb [options] <file>"

  parser.on("-n", "--numbers", "Display line numbers") do
    opts[:show_line_numbers] = true
  end

  parser.on("-sINDEX", "--stream INDEX", Integer, "Display only the specified stream index") do |index|
    opts[:stream_index] = index
  end

  parser.on("-wWIDTH", "--width WIDTH", Integer, "Specify column width") do |width|
    opts[:column_width] = width
  end

  parser.on("-h", "--head", "Treat the first line as a header") do
    opts[:treat_as_header] = true
  end

  parser.on("-lLINE", "--line LINE", Integer, "Display only the specified line") do |line|
    opts[:specific_line] = line
  end

  parser.on_tail("--help", "Show this message") do
    puts parser
    exit
  end
end.parse!

# Input file
fname = ARGV.first
if fname.nil?
  STDERR.puts "Error: No input file specified."
  exit 1
end

begin
  txt = File.open(fname).read.ascii!
rescue
  STDERR.puts "Error: Can't open file #{fname}"
  exit 1
end

# Parse markers
atxt = txt.bytes.clone
idx = 0
markers = []

begin
  markpos = atxt.index(0xFF)
  break if markpos.nil?
  nextbyte = atxt[markpos + 1]

  case nextbyte
  when NSM.bytes.last
    markers << [NSM.bytes.last, idx + markpos + 1]
  when NFM.bytes.last
    markers << [NFM.bytes.last, idx + markpos + 1]
  when NCM.bytes.last
    markers << [NCM.bytes.last, idx + markpos + 1]
  else
    STDERR.puts "Error: Invalid XTXT marker 0x#{nextbyte.to_s(16)} at #{markpos}"
    exit 1
  end
  idx = markpos + idx + 2
  atxt = atxt.slice(markpos + 2, atxt.size)
end until atxt.empty?

if markers.empty?
  markers << [-1, atxt.size + 1]
end

streams = []
streamwidths = []
frameno = 0
streamno = 0
idx = 0

markers.each do |mark|
  mtype = mark.first
  mslice = mark.last
  aline = txt.bytes[idx..mslice - 2]

  aline = aline.empty? ? "".ascii! : aline.pack("c*").ascii!

  case mtype
  when -1 # Plain text file, single stream
    streams[0] = aline.split("\n")
    longest = streams[0].map { |line| line.bytes.size }.max
    streamwidths[0] = longest
    break

  when NSM.bytes.last
    streams[streamno] ||= []
    uline = aline.clone.utf8!

    if uline.bytes[0..2] == [0xEF, 0xBB, 0xBF]
      uline = uline[1..-1] # Remove BOM
    end

    streams[streamno][frameno] = uline
    streamwidths[streamno] = [streamwidths[streamno].to_i, uline.size].max
    streamno += 1
    idx += 2

  when NFM.bytes.last
    idx += 2
    streamno = 0
    frameno += 1

  else
    STDERR.puts "Invalid XTXT sequence at index #{idx}"
    exit 1
  end
  idx += aline.bytes.size
end

# Output the streams
streamlengths = streams.map(&:size)
longest_stream = streamlengths.max
linenochars = [longest_stream.to_s.size, 3].max
linenofmt = "%0#{linenochars}d"

0.upto(longest_stream - 1) do |fno|
  next if opts[:specific_line] && fno + 1 != opts[:specific_line]

  line = ""
  0.upto(streams.size - 1) do |streamidx|
    next if opts[:stream_index] && streamidx != opts[:stream_index]

    columnsize = streamwidths[streamidx]
    column = streams[streamidx][fno] || ""
    column = column.chomp.utf8!
    line << column.ljust(columnsize)[0..columnsize - 1]
  end

  if opts[:show_line_numbers]
    puts "#{linenofmt % (fno + 1)} #{line}"
  else
    puts line
  end
end
